internal/handler/common/update_array_operators.go

```diff
@@ -120,366 +103,353 @@ func checkUnsuitableValueError(doc *types.Document, key string, path types.Path)
-			each.Append(addToSetValueRaw)
-		}
+	value, err := doc.GetByPath(path)
+	if err != nil {
+		return false, err
+	}
 
-		for i := 0; i < each.Len(); i++ {
-			value := must.NotFail(each.Get(i))
+	array, ok := value.(*types.Array)
+	if !ok {
+		return false, NewUpdateError(
+			handlererrors.ErrBadValue,
+			fmt.Sprintf(
+				"The field '%s' must be an array but is of type '%s' in document {_id: %s}",
+				key, handlerparams.AliasFromType(value), types.FormatAnyValue(must.NotFail(doc.Get("_id"))),
+			),
+			command,
+		)
+	}
 
-			if array.Contains(value) {
-				continue
-			}
+	if each == nil {
+		each = types.MakeArray(1)
+		each.Append(setVal)
+	}
 
-			changed = true
+	var changed bool
 
-			array.Append(value)
-		}
+	for i := range each.Len() {
+		elem := must.NotFail(each.Get(i))
 
-		if err = doc.SetByPath(path, array); err != nil {
-			return false, lazyerrors.Error(err)
+		if array.Contains(elem) {
+			continue
 		}
+
+		changed = true
+
+		array.Append(elem)
+	}
+
+	if err = doc.SetByPath(path, array); err != nil {
+		return false, lazyerrors.Error(err)
 	}
 
 	return changed, nil
 }
 
 // processPullAllArrayUpdateExpression changes document according to $pullAll array update operator.
 // If the document was changed it returns true.
-func processPullAllArrayUpdateExpression(doc, update *types.Document) (bool, error) {
-	var changed bool
-
-	iter := update.Iterator()
-	defer iter.Close()
-
-	for {
-		key, pullAllValueRaw, err := iter.Next()
-		if err != nil {
-			if errors.Is(err, iterator.ErrIteratorDone) {
-				break
-			}
-
-			return false, lazyerrors.Error(err)
-		}
-
-		path, err := types.NewPathFromString(key)
-		if err != nil {
-			return false, lazyerrors.Error(err)
-		}
-
-		// If the path does not exist, create a new array and set it.
-		if !doc.HasByPath(path) {
-			if err = doc.SetByPath(path, types.MakeArray(1)); err != nil {
-				return false, handlererrors.NewWriteErrorMsg(
-					handlererrors.ErrUnsuitableValueType,
-					err.Error(),
-				)
-			}
-		}
+func processPullAllArrayUpdateExpression(command string, doc *types.Document, key string, pullVal any) (bool, error) {
+	pullArray, ok := pullVal.(*types.Array)
+	if !ok {
+		return false, NewUpdateError(
+			handlererrors.ErrBadValue,
+			fmt.Sprintf(
+				"The field '%s' must be an array but is of type '%s'",
+				key, handlerparams.AliasFromType(pullVal),
+			),
+			command,
+		)
+	}
 
-		val, err := doc.GetByPath(path)
-		if err != nil {
-			return false, lazyerrors.Error(err)
-		}
+	path, err := types.NewPathFromString(key)
+	if err != nil {
+		return false, lazyerrors.Error(err)
+	}
 
-		array, ok := val.(*types.Array)
-		if !ok {
-			return false, handlererrors.NewWriteErrorMsg(
-				handlererrors.ErrBadValue,
-				fmt.Sprintf(
-					"The field '%s' must be an array but is of type '%s' in document {_id: %s}",
-					key, handlerparams.AliasFromType(val), types.FormatAnyValue(must.NotFail(doc.Get("_id"))),
-				),
-			)
+	if !doc.HasByPath(path) {
```

Comment:
Should it create a new array and set it? The logic appears to have changed.

Reply:
Yes, the logic is changed. MongoDB doesn't add an array for non-existent fields in case of `$pop`, `$pull` and `$pullAll` operators.

We had this compatability bug before and it was not visible in our tests.
1. As per the previous logic (in all three operators), we will add an array for non-existent fields to the `doc`, but the `changed` flag will be `false` when function returns :upside_down_face:
2. In case an existing document, there won't be any compat issue, because we won't update the DB backends if `changed` is false.
3. But in case of upsert, where we ignore the `changed` flag and always push contents of `doc` pointer to DB backends, we will face the issue.
4. We didn't have a test case for [upsert + non-existent field] scenario, so this issue was not visible to us.

I have added compat test cases (`NonExistentFieldUpsert`) for all three operators and also modified `testUpdateCompat()` to compare the inserted documents, in case of upsert. :cowboy_hat_face: